// @lenovo-software/lsa-clients-common/dist/defaults/defaultCryptoSubtle.js
"use strict";
// Copyright © 2023 Lenovo. All rights reserved.
// Confidential and Restricted
Object.defineProperty(exports, "__esModule", { value: true });
exports.DefaultCryptoSubtle = void 0;
const webcrypto_1 = require("@peculiar/webcrypto");
class DefaultCryptoSubtle {
    constructor() {
        this._crypto = new webcrypto_1.Crypto();
    }
    generateKey(obj, exportable, a) {
        return this._crypto.subtle.generateKey(obj, exportable, a);
    }
    exportKey(str, k) {
        return this._crypto.subtle.exportKey(str, k);
    }
    importKey(str, k, obj, exportable, a) {
        return this._crypto.subtle.importKey(str, k, obj, exportable, a);
    }
    sign(obj, k, payload) {
        return this._crypto.subtle.sign(obj, k, payload);
    }
    verify(obj, k, signature, payload) {
        return this._crypto.subtle.verify(obj, k, signature, payload);
    }
    encrypt(algorithm, k, data) {
        return this._crypto.subtle.encrypt(algorithm, k, data);
    }
    decrypt(algorithm, k, data) {
        return this._crypto.subtle.decrypt(algorithm, k, data);
    }
    getRandomValues(array) {
        return this._crypto.getRandomValues(array);
    }
}
exports.DefaultCryptoSubtle = DefaultCryptoSubtle;
