// @lenovo-software/lsa-clients-common/dist/js/interop/LStudentInterop.js
"use strict";
// Copyright © 2023 Lenovo. All rights reserved.
// Confidential and Restricted
Object.defineProperty(exports, "__esModule", { value: true });
exports.LStudentInterop = void 0;
const internal_1 = require("../../internal");
class LStudentInterop {
    constructor() {
        this.lsaClient = null;
    }
    get policy() {
        if (internal_1.LSAClient) {
            this.lsaClient = internal_1.LSAClient.getInstance();
            return {
                api_server: this.lsaClient.clientParams.apiServer,
                provisioning_code: (this.lsaClient.clientParams.hotProvisioningCode.length > 0) ?
                    this.lsaClient.clientParams.hotProvisioningCode :
                    this.lsaClient.clientParams.provisioningCode
            };
        }
        return undefined;
    }
    get apiServer() {
        if (internal_1.LSAClient) {
            this.lsaClient = internal_1.LSAClient.getInstance();
            return this.lsaClient.clientParams.apiServerForHTTPS;
        }
        return undefined;
    }
}
exports.LStudentInterop = LStudentInterop;
