// @lenovo-software/lsa-clients-common/dist/js/interop/windowInterop.js
"use strict";
// Copyright © 2023 Lenovo. All rights reserved.
// Confidential and Restricted
Object.defineProperty(exports, "__esModule", { value: true });
exports.window = void 0;
const internal_1 = require("../../internal");
var window = {
    crypto: {
        subtle: {
            generateKey: (obj, exportable, a) => {
                return internal_1.LSAClient.getInstance().cryptoSubtleCallback.generateKey(obj, exportable, a);
            },
            exportKey: (str, k) => {
                return internal_1.LSAClient.getInstance().cryptoSubtleCallback.exportKey(str, k);
            },
            importKey: (str, k, obj, exportable, a) => {
                return internal_1.LSAClient.getInstance().cryptoSubtleCallback.importKey(str, k, obj, exportable, a);
            },
            sign: (obj, k, payload) => {
                return internal_1.LSAClient.getInstance().cryptoSubtleCallback.sign(obj, k, payload);
            },
            verify: (obj, k, signature, payload) => {
                return internal_1.LSAClient.getInstance().cryptoSubtleCallback.verify(obj, k, signature, payload);
            }
        }
    }
};
exports.window = window;
