// @lenovo-software/lsa-clients-common/dist/js/defs.js
"use strict";
// Copyright © 2023 Lenovo. All rights reserved.
// Confidential and Restricted
Object.defineProperty(exports, "__esModule", { value: true });
exports.STUDENT_CONFERENCE_TYPE_VIDEO = exports.TEACHER_CONFERENCE_TYPE_AUDIO = exports.TEACHER_CONFERENCE_TYPE_VIDEO = exports.TEACHER_CONFERENCE_TYPE_NONE = exports.SyncStorageThrottleMax = exports.SyncStorageThrottleBackOffAmount = exports.SyncStorageDefaultInterval = exports.SyncStorageMaxErrorsBeforeThrottle = exports.MetricReporterReportingEndpoint = exports.MetricReporterConfigEndpoint = exports.MetricReporterCheckInterval = exports.MetricReporterMinimumReportingInterval = exports.LogBufferSize = exports.CatchAllOrg = exports.WebHelperExtensionIds = exports.RouteEnum = exports.ConferenceStateEnum = exports.StatusPageIcons = exports.LEDColorsString = exports.LEDColors = exports.SigningRequirementEnum = exports.ConnectionStateEnum = void 0;
/* eslint-disable no-unused-vars */
var ConnectionStateEnum = Object.freeze({
    "Connected": 1,
    "Disconnected": 2,
    "Trying": 3
});
exports.ConnectionStateEnum = ConnectionStateEnum;
var SigningRequirementEnum = Object.freeze({
    "Enforced": 1,
    "None": 2,
    "Unknown": 3
});
exports.SigningRequirementEnum = SigningRequirementEnum;
var LEDColors = Object.freeze({
    "Green": 1,
    "Amber": 2,
    "Red": 3
});
exports.LEDColors = LEDColors;
var LEDColorsString = Object.freeze({
    Green: 'Green',
    Amber: 'Amber',
    Red: 'Red'
});
exports.LEDColorsString = LEDColorsString;
var StatusPageIcons = Object.freeze({
    "Good": 1,
    "Warn": 2,
    "Error": 3,
    "Info": 4
});
exports.StatusPageIcons = StatusPageIcons;
var ConferenceStateEnum = Object.freeze({
    "Disconnected": 1,
    "InProcess": 2,
    "Connected": 3
});
exports.ConferenceStateEnum = ConferenceStateEnum;
let RouteEnum = Object.freeze({
    "Chat": "build/index.html#/chat",
    "Speaker": "build/index.html#/speaker-audio"
});
exports.RouteEnum = RouteEnum;
var WebHelperExtensionIds = [
    "baleiojnjpgeojohhhfbichcodgljmnj" // LanSchool Air web helper (production)
    ,
    "ljejjhdjfagbhnehmnhihhkcbanajcoa" // LanSchool Air web helper (beta)
    ,
    "mmpgfeapmofhobeagaffmjfnfaagjkbk" // LanSchool Air web helper (beta 2)
    ,
    "omkignahognlejgkcbcmcpafacamikmp" // LanSchool Air web helper (beta 3)
    ,
    "keknjhjnninjadlkapachhhjfmfnofcb" // Self-hosted extension
];
exports.WebHelperExtensionIds = WebHelperExtensionIds;
var CatchAllOrg = 'deadbeef-4811-30f6-42e7-7d8421b2bece';
exports.CatchAllOrg = CatchAllOrg;
var LogBufferSize = 10000;
exports.LogBufferSize = LogBufferSize;
var MetricReporterMinimumReportingInterval = 5000;
exports.MetricReporterMinimumReportingInterval = MetricReporterMinimumReportingInterval;
var MetricReporterCheckInterval = 300000;
exports.MetricReporterCheckInterval = MetricReporterCheckInterval;
var MetricReporterConfigEndpoint = '0/lsa/common/metrics/config';
exports.MetricReporterConfigEndpoint = MetricReporterConfigEndpoint;
var MetricReporterReportingEndpoint = '0/lsa/common/metrics/events';
exports.MetricReporterReportingEndpoint = MetricReporterReportingEndpoint;
var SyncStorageMaxErrorsBeforeThrottle = 5;
exports.SyncStorageMaxErrorsBeforeThrottle = SyncStorageMaxErrorsBeforeThrottle;
var SyncStorageDefaultInterval = 1000;
exports.SyncStorageDefaultInterval = SyncStorageDefaultInterval;
var SyncStorageThrottleBackOffAmount = 5000;
exports.SyncStorageThrottleBackOffAmount = SyncStorageThrottleBackOffAmount;
var SyncStorageThrottleMax = 120000;
exports.SyncStorageThrottleMax = SyncStorageThrottleMax;
// Bitmaskable conference types
const TEACHER_CONFERENCE_TYPE_NONE = 0;
exports.TEACHER_CONFERENCE_TYPE_NONE = TEACHER_CONFERENCE_TYPE_NONE;
const TEACHER_CONFERENCE_TYPE_VIDEO = 1;
exports.TEACHER_CONFERENCE_TYPE_VIDEO = TEACHER_CONFERENCE_TYPE_VIDEO;
const TEACHER_CONFERENCE_TYPE_AUDIO = 2;
exports.TEACHER_CONFERENCE_TYPE_AUDIO = TEACHER_CONFERENCE_TYPE_AUDIO;
const STUDENT_CONFERENCE_TYPE_VIDEO = 4;
exports.STUDENT_CONFERENCE_TYPE_VIDEO = STUDENT_CONFERENCE_TYPE_VIDEO;
if (typeof exports !== 'undefined') {
    exports.ConnectionStateEnum = ConnectionStateEnum;
    exports.SigningRequirementEnum = SigningRequirementEnum;
    exports.WebHelperExtensionIds = WebHelperExtensionIds;
    exports.LEDColors = LEDColors;
    exports.ConferenceStateEnum = ConferenceStateEnum;
    exports.RouteEnum = RouteEnum;
    exports.CatchAllOrg = CatchAllOrg;
    exports.MetricReporterMinimumReportingInterval = MetricReporterMinimumReportingInterval;
    exports.MetricReporterCheckInterval = MetricReporterCheckInterval;
    exports.MetricReporterConfigEndpoint = MetricReporterConfigEndpoint;
    exports.MetricReporterReportingEndpoint = MetricReporterReportingEndpoint;
    exports.SyncStorageMaxErrorsBeforeThrottle = SyncStorageMaxErrorsBeforeThrottle;
    exports.SyncStorageDefaultInterval = SyncStorageDefaultInterval;
    exports.SyncStorageThrottleBackOffAmount = SyncStorageThrottleBackOffAmount;
    exports.SyncStorageThrottleMax = SyncStorageThrottleMax;
    exports.LogBufferSize = LogBufferSize;
}
