// @lenovo-software/lsa-clients-common/dist/models/IntelliTaskInfoModel.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LastIntellitaskMessageState = exports.SimpleTabState = void 0;
const lodash_1 = require("lodash");
const internal_1 = require("../internal");
class SimpleTabState {
    /**
     * Map OtmBrowserTab to tab state which can be used to identify any change in the tab state
     */
    static from(tab) {
        const tabState = (0, lodash_1.pick)(tab, ['title', 'url', 'isActive', 'audible', 'muted', 'audioElements', 'usedDataGathering']);
        tabState.hasContent = !!tab.content;
        return tabState;
    }
}
exports.SimpleTabState = SimpleTabState;
class LastIntellitaskMessageState {
    static from(otmBrowserTabs) {
        var _a, _b;
        const tabs = (_a = otmBrowserTabs === null || otmBrowserTabs === void 0 ? void 0 : otmBrowserTabs.tabs) !== null && _a !== void 0 ? _a : [];
        return {
            tabs: tabs.map(SimpleTabState.from),
            activeTabInfo: otmBrowserTabs === null || otmBrowserTabs === void 0 ? void 0 : otmBrowserTabs.activeTabInfo,
            isWebExtensionAvailable: !!((_b = internal_1.store.getState().webExtension) === null || _b === void 0 ? void 0 : _b.isConnected)
        };
    }
}
exports.LastIntellitaskMessageState = LastIntellitaskMessageState;
