// @lenovo-software/lsa-clients-common/dist/models/limitingModels.js
"use strict";
// Copyright © 2023 Lenovo. All rights reserved.
// Confidential and Restricted
Object.defineProperty(exports, "__esModule", { value: true });
exports.AppLimitParams = exports.AppLimitBlockType = exports.WebLimitParams = exports.WebLimitBlockType = exports.BlankScreenParameters = void 0;
class BlankScreenParameters {
    constructor() {
        this.backgroundUrl = '';
        this.className = '';
        this.shouldBlockInput = false;
    }
}
exports.BlankScreenParameters = BlankScreenParameters;
var WebLimitBlockType;
(function (WebLimitBlockType) {
    WebLimitBlockType[WebLimitBlockType["None"] = 0] = "None";
    WebLimitBlockType[WebLimitBlockType["BlockAll"] = 1] = "BlockAll";
    WebLimitBlockType[WebLimitBlockType["Block"] = 2] = "Block";
    WebLimitBlockType[WebLimitBlockType["Allow"] = 3] = "Allow";
})(WebLimitBlockType = exports.WebLimitBlockType || (exports.WebLimitBlockType = {}));
class WebLimitParams {
    constructor() {
        this.className = '';
    }
}
exports.WebLimitParams = WebLimitParams;
var AppLimitBlockType;
(function (AppLimitBlockType) {
    AppLimitBlockType[AppLimitBlockType["None"] = 0] = "None";
    AppLimitBlockType[AppLimitBlockType["Block"] = 1] = "Block";
    AppLimitBlockType[AppLimitBlockType["Allow"] = 2] = "Allow";
})(AppLimitBlockType = exports.AppLimitBlockType || (exports.AppLimitBlockType = {}));
class AppLimitParams {
    constructor() {
        this.className = '';
    }
}
exports.AppLimitParams = AppLimitParams;
