// @lenovo-software/lsa-clients-common/dist/models/meetingModels.js
"use strict";
// Copyright © 2024 Lenovo. All rights reserved.
// Confidential and Restricted
Object.defineProperty(exports, "__esModule", { value: true });
exports.MeetingDirectiveModel = exports.MeetingTypeEnum = void 0;
const _TeacherMeetingTypeNone = require('../js/defs').TEACHER_CONFERENCE_TYPE_NONE;
const _TeacherMeetingTypeAudio = require('../js/defs').TEACHER_CONFERENCE_TYPE_AUDIO;
const _TeacherMeetingTypeVideo = require('../js/defs').TEACHER_CONFERENCE_TYPE_VIDEO;
const _StudentMeetingTypeVideo = require('../js/defs').STUDENT_CONFERENCE_TYPE_VIDEO;
var MeetingTypeEnum;
(function (MeetingTypeEnum) {
    MeetingTypeEnum[MeetingTypeEnum["TeacherMeetingTypeNone"] = _TeacherMeetingTypeNone] = "TeacherMeetingTypeNone";
    MeetingTypeEnum[MeetingTypeEnum["TeacherMeetingTypeAudio"] = _TeacherMeetingTypeAudio] = "TeacherMeetingTypeAudio";
    MeetingTypeEnum[MeetingTypeEnum["TeacherMeetingTypeVideo"] = _TeacherMeetingTypeVideo] = "TeacherMeetingTypeVideo";
    MeetingTypeEnum[MeetingTypeEnum["StudentMeetingTypeVideo"] = _StudentMeetingTypeVideo] = "StudentMeetingTypeVideo";
})(MeetingTypeEnum = exports.MeetingTypeEnum || (exports.MeetingTypeEnum = {}));
const _defaultMeetingTypeValue = MeetingTypeEnum.TeacherMeetingTypeNone;
const _defaultWindowed = false;
class MeetingDirectiveModel {
    constructor() {
        this.meetingId = '';
        this.meetingType = _defaultMeetingTypeValue;
        this.windowed = _defaultWindowed;
        this.apiServer = '';
        this.classId = '';
        this.oauthToken = '';
    }
    static fromAny(data) {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
        if (!data) {
            return new MeetingDirectiveModel();
        }
        if (data instanceof MeetingDirectiveModel) {
            return Object.assign({}, data);
        }
        const ret = new MeetingDirectiveModel();
        ret.meetingId = (_b = (_a = data.meetingConfig) === null || _a === void 0 ? void 0 : _a.meetingId) !== null && _b !== void 0 ? _b : null;
        ret.meetingType = (_d = (_c = data.conferenceData) === null || _c === void 0 ? void 0 : _c.conferenceType) !== null && _d !== void 0 ? _d : _defaultMeetingTypeValue;
        ret.windowed = (_f = (_e = data.meetingConfig) === null || _e === void 0 ? void 0 : _e.windowed) !== null && _f !== void 0 ? _f : _defaultWindowed;
        ret.apiServer = (_g = data.apiServer) !== null && _g !== void 0 ? _g : '';
        ret.classId = (_h = data.classId) !== null && _h !== void 0 ? _h : '';
        ret.oauthToken = (_k = (_j = data.conferenceData) === null || _j === void 0 ? void 0 : _j.oauthToken) !== null && _k !== void 0 ? _k : '';
        return ret;
    }
}
exports.MeetingDirectiveModel = MeetingDirectiveModel;
