// @lenovo-software/lsa-clients-common/dist/models/provisioningModels.js
"use strict";
// Copyright © 2023 Lenovo. All rights reserved.
// Confidential and Restricted
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProvisioningParams = exports.ProvisioningData = void 0;
class ProvisioningData {
    constructor() {
        this.deviceID = '';
        this.deviceId = '';
        this.id = '';
        this.orgId = '';
        this.orgID = '';
    }
    static fromAny(data) {
        var _a, _b, _c, _d, _e;
        if (!data) {
            return new ProvisioningData();
        }
        if (data instanceof ProvisioningData) {
            return data;
        }
        const ret = new ProvisioningData();
        ret.deviceId = ret.deviceID = (_b = (_a = data.deviceId) !== null && _a !== void 0 ? _a : data.deviceID) !== null && _b !== void 0 ? _b : '';
        ret.id = (_c = data.id) !== null && _c !== void 0 ? _c : '';
        ret.orgId = ret.orgID = (_e = (_d = data.orgId) !== null && _d !== void 0 ? _d : data.orgID) !== null && _e !== void 0 ? _e : '';
        return ret;
    }
}
exports.ProvisioningData = ProvisioningData;
class ProvisioningParams {
    constructor() {
        this.deviceName = '';
        this.machineCode = '';
        this.hotProvisioningCode = '';
        this.provisioningCode = '';
        this.emailAddr = '';
        this.deviceId = '';
        this.osType = '';
        this.deviceType = '';
    }
    static fromAny(data) {
        var _a, _b, _c, _d, _e, _f, _g, _h;
        if (!data) {
            return new ProvisioningParams();
        }
        if (data instanceof ProvisioningParams) {
            return data;
        }
        const ret = new ProvisioningParams();
        ret.deviceName = (_a = data.deviceName) !== null && _a !== void 0 ? _a : '';
        ret.machineCode = (_b = data.machineCode) !== null && _b !== void 0 ? _b : '';
        ret.provisioningCode = (_c = data.provisioningCode) !== null && _c !== void 0 ? _c : '';
        ret.hotProvisioningCode = (_d = data.hotProvisioningCode) !== null && _d !== void 0 ? _d : '';
        ret.emailAddr = (_e = data.emailAddr) !== null && _e !== void 0 ? _e : '';
        ret.deviceId = (_f = data.deviceId) !== null && _f !== void 0 ? _f : '';
        ret.osType = (_g = data.osType) !== null && _g !== void 0 ? _g : '';
        ret.deviceType = (_h = data.deviceType) !== null && _h !== void 0 ? _h : '';
        return ret;
    }
}
exports.ProvisioningParams = ProvisioningParams;
