// @lenovo-software/lsa-clients-common/dist/models/socketMessageModels.js
"use strict";
// Copyright © 2023 Lenovo. All rights reserved.
// Confidential and Restricted
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClearRaiseHandMessage = exports.RaiseHandMessage = exports.CurrentScreenMessage = void 0;
class CurrentScreenMessage {
    constructor() {
        this.type = 'currentScreen';
        this.currentScreenMessage = {
            size: {
                width: 0,
                height: 0
            },
            image: {
                data: '',
                type: ''
            },
            activeTabFailure: false
        };
        this.destination = '';
        this.socketId = '';
    }
}
exports.CurrentScreenMessage = CurrentScreenMessage;
class RaiseHandMessage {
    constructor() {
        this.type = 'raiseHand';
        this.raiseHandMessage = {
            sendTime: new Date(),
            loginName: ''
        };
        this.destination = '';
        this.socketId = '';
    }
}
exports.RaiseHandMessage = RaiseHandMessage;
class ClearRaiseHandMessage {
    constructor() {
        this.type = 'clearRaiseHand';
        this.clearRaiseHandMessage = {
            sendTime: new Date(),
            loginName: ''
        };
        this.destination = '';
        this.socketId = '';
    }
}
exports.ClearRaiseHandMessage = ClearRaiseHandMessage;
