// @lenovo-software/lsa-clients-common/dist/models/teacherMessages.js
"use strict";
// Copyright © 2023 Lenovo. All rights reserved.
// Confidential and Restricted
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClientState = exports.TeacherInfo = exports.LimitStateMessage = exports.Desktop = exports.Sound = exports.Printing = exports.Applications = exports.Websites = void 0;
class Websites {
    constructor() {
        this.allowed = [];
        this.blocked = [];
        this.limit = false;
    }
}
exports.Websites = Websites;
class Applications {
    constructor() {
        this.allowed = [];
        this.blocked = [];
        this.limit = false;
    }
}
exports.Applications = Applications;
class Printing {
    constructor() {
        this.limit = false;
    }
}
exports.Printing = Printing;
class Sound {
    constructor() {
        this.limit = false;
    }
}
exports.Sound = Sound;
class Desktop {
    constructor() {
        this.limit = false;
        this.message = '';
        this.kioskMode = true;
        this.shouldBlockInput = false;
    }
}
exports.Desktop = Desktop;
class LimitStateMessage {
    constructor() {
        this.websites = new Websites();
        this.applications = new Applications();
        this.printing = new Printing();
        this.sound = new Sound();
        this.desktop = new Desktop();
    }
}
exports.LimitStateMessage = LimitStateMessage;
class TeacherInfo {
    constructor() {
        this.username = 'Frank Sanchez';
        this.guid = 'Frank-Guid-Sanchez';
    }
}
exports.TeacherInfo = TeacherInfo;
class ClientState {
    constructor() {
        this.type = 'limitState';
        this.limitStateMessage = new LimitStateMessage();
        this.teacherInfo = new TeacherInfo();
    }
}
exports.ClientState = ClientState;
