// @lenovo-software/lsa-clients-common/dist/models/thumbnail.js
"use strict";
// Copyright © 2023 Lenovo. All rights reserved.
// Confidential and Restricted
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThumbnailRequest = exports.ThumbnailImage = exports.ThumbnailSize = exports.ThumbnailSource = exports.ThumbnailFormat = exports.DEFAULT_THUMBNAIL_HEIGHT = exports.DEFAULT_THUMBNAIL_WIDTH = void 0;
exports.DEFAULT_THUMBNAIL_WIDTH = 480;
exports.DEFAULT_THUMBNAIL_HEIGHT = 240;
var ThumbnailFormat;
(function (ThumbnailFormat) {
    ThumbnailFormat[ThumbnailFormat["JPG"] = 0] = "JPG";
    ThumbnailFormat[ThumbnailFormat["PNG"] = 1] = "PNG";
})(ThumbnailFormat = exports.ThumbnailFormat || (exports.ThumbnailFormat = {}));
var ThumbnailSource;
(function (ThumbnailSource) {
    ThumbnailSource[ThumbnailSource["ActiveTab"] = 1] = "ActiveTab";
    ThumbnailSource[ThumbnailSource["FullScreen"] = 2] = "FullScreen";
    ThumbnailSource[ThumbnailSource["Unknown"] = 3] = "Unknown";
})(ThumbnailSource = exports.ThumbnailSource || (exports.ThumbnailSource = {}));
class ThumbnailSize {
    constructor() {
        this.width = exports.DEFAULT_THUMBNAIL_WIDTH;
        this.height = exports.DEFAULT_THUMBNAIL_HEIGHT;
    }
}
exports.ThumbnailSize = ThumbnailSize;
// What the caller returns to the client and what the student sends back to the teacher
class ThumbnailImage {
    constructor() {
        this.image = '';
        this.type = 'base64';
    }
}
exports.ThumbnailImage = ThumbnailImage;
// The data structure providing the thumbnail parameters as requested by the teacher
class ThumbnailRequest {
}
exports.ThumbnailRequest = ThumbnailRequest;
