// @lenovo-software/lsa-clients-common/dist/internal.js
"use strict";
/**
 * Here we need to maintain dependency order
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.httpAPI = exports.StatusController = exports.routeVerifier = exports.PollingTask = exports.CallbackQueue = exports.SoftBase64Exchange = exports.Base64ToArrayBufferEx = exports.ArrayBufferToBase64ExPadding = exports.ArrayBufferToBase64Ex = exports.logger = exports.LogClientEx = exports.window = exports.UserInfoRetriever = exports.showMonitoringDialog = exports.chromeMessenger = exports.chrome = void 0;
/** interop */
__exportStar(require("./js/interop/loggerInterop"), exports);
var chromeInterop_1 = require("./js/interop/chromeInterop");
Object.defineProperty(exports, "chrome", { enumerable: true, get: function () { return chromeInterop_1.chrome; } });
Object.defineProperty(exports, "chromeMessenger", { enumerable: true, get: function () { return chromeInterop_1.chromeMessenger; } });
__exportStar(require("./js/interop/conferenceControllerInterop"), exports);
__exportStar(require("./js/interop/httpRequest"), exports);
__exportStar(require("./js/interop/telemetryInterop"), exports);
__exportStar(require("./js/interop/LStudentInterop"), exports);
var orgVerificationControllerInterop_1 = require("./js/interop/orgVerificationControllerInterop");
Object.defineProperty(exports, "showMonitoringDialog", { enumerable: true, get: function () { return orgVerificationControllerInterop_1.showMonitoringDialog; } });
__exportStar(require("./js/interop/pkiInterop"), exports);
__exportStar(require("./js/interop/socketControllerInterop"), exports);
__exportStar(require("./js/interop/statusControllerInterop"), exports);
__exportStar(require("./js/interop/storageInterop"), exports);
var userInfo_1 = require("./js/interop/userInfo");
Object.defineProperty(exports, "UserInfoRetriever", { enumerable: true, get: function () { return userInfo_1.UserInfoRetriever; } });
var windowInterop_1 = require("./js/interop/windowInterop");
Object.defineProperty(exports, "window", { enumerable: true, get: function () { return windowInterop_1.window; } });
/** logging */
var logClient_1 = require("./js/logging/logClient");
Object.defineProperty(exports, "LogClientEx", { enumerable: true, get: function () { return logClient_1.LogClientEx; } });
var logger_1 = require("./js/logging/logger");
Object.defineProperty(exports, "logger", { enumerable: true, get: function () { return logger_1.logger; } });
/** i18n */
__exportStar(require("./i18n/configs/config"), exports);
__exportStar(require("./i18n/configs/i18next.config"), exports);
__exportStar(require("./i18n/controller/i18nController"), exports);
__exportStar(require("./i18n/mapper/locale.mapper"), exports);
/** defaults */
__exportStar(require("./defaults/defaultCryptoSubtle"), exports);
__exportStar(require("./defaults/defaultLogger"), exports);
__exportStar(require("./defaults/defaultNetworkCallback"), exports);
__exportStar(require("./defaults/defaultOrgCheck"), exports);
__exportStar(require("./defaults/defaultRuntime"), exports);
__exportStar(require("./defaults/defaultStorage"), exports);
__exportStar(require("./defaults/defaultThumbnail"), exports);
/** interfaces */
__exportStar(require("./interfaces/batteryInstrumentation"), exports);
__exportStar(require("./interfaces/browserInstrumentation"), exports);
__exportStar(require("./interfaces/conferenceDirectives"), exports);
__exportStar(require("./interfaces/contextMenu"), exports);
__exportStar(require("./interfaces/cryptoSubtle"), exports);
__exportStar(require("./interfaces/event"), exports);
__exportStar(require("./interfaces/knownClientStatus"), exports);
__exportStar(require("./interfaces/limitingHandler"), exports);
__exportStar(require("./interfaces/localeHandler"), exports);
__exportStar(require("./interfaces/logger"), exports);
__exportStar(require("./interfaces/meetingDirective"), exports);
__exportStar(require("./interfaces/networkCallback"), exports);
__exportStar(require("./interfaces/runningAppsInstrumentation"), exports);
__exportStar(require("./interfaces/runtime"), exports);
__exportStar(require("./interfaces/storage"), exports);
__exportStar(require("./interfaces/studentScreenShareController"), exports);
__exportStar(require("./interfaces/thumbnailInstrumentation"), exports);
__exportStar(require("./interfaces/ui"), exports);
__exportStar(require("./interfaces/unverifiedOrgOption"), exports);
__exportStar(require("./interfaces/webHistory"), exports);
__exportStar(require("./models/screenShare"), exports);
/** models */
__exportStar(require("./models/batteryStatus"), exports);
__exportStar(require("./models/browserTabs"), exports);
__exportStar(require("./models/chat"), exports);
__exportStar(require("./models/clientParams"), exports);
__exportStar(require("./models/conferencingModels"), exports);
__exportStar(require("./models/currentWebsite"), exports);
__exportStar(require("./models/eventModels"), exports);
__exportStar(require("./models/limitingModels"), exports);
__exportStar(require("./models/meetingModels"), exports);
__exportStar(require("./models/provisioningModels"), exports);
__exportStar(require("./models/runningApp"), exports);
__exportStar(require("./models/socketMessageModels"), exports);
__exportStar(require("./models/studentStatusModels"), exports);
__exportStar(require("./models/teacherMessages"), exports);
__exportStar(require("./models/telemetryModels"), exports);
__exportStar(require("./models/thumbnail"), exports);
__exportStar(require("./models/token"), exports);
__exportStar(require("./models/webHistory"), exports);
__exportStar(require("./models/IntelliTaskInfoModel"), exports);
/** ui */
__exportStar(require("./ui/deadOrgNotificationDialog"), exports);
/** Shared */
__exportStar(require("./shared/commonWindow"), exports);
/**
 * js
 * NOTE: .JS files requires names export
 */
var base64_1 = require("./js/base64");
Object.defineProperty(exports, "ArrayBufferToBase64Ex", { enumerable: true, get: function () { return base64_1.ArrayBufferToBase64Ex; } });
Object.defineProperty(exports, "ArrayBufferToBase64ExPadding", { enumerable: true, get: function () { return base64_1.ArrayBufferToBase64ExPadding; } });
Object.defineProperty(exports, "Base64ToArrayBufferEx", { enumerable: true, get: function () { return base64_1.Base64ToArrayBufferEx; } });
Object.defineProperty(exports, "SoftBase64Exchange", { enumerable: true, get: function () { return base64_1.SoftBase64Exchange; } });
var cbq_1 = require("./js/cbq");
Object.defineProperty(exports, "CallbackQueue", { enumerable: true, get: function () { return cbq_1.CallbackQueue; } });
__exportStar(require("./js/metrics"), exports);
var pollingTask_1 = require("./js/pollingTask");
Object.defineProperty(exports, "PollingTask", { enumerable: true, get: function () { return pollingTask_1.PollingTask; } });
var routeVerifier_1 = require("./js/routeVerifier");
Object.defineProperty(exports, "routeVerifier", { enumerable: true, get: function () { return routeVerifier_1.routeVerifier; } });
var statusController_1 = require("./js/statusController");
Object.defineProperty(exports, "StatusController", { enumerable: true, get: function () { return statusController_1.StatusController; } });
/** root files */
exports.httpAPI = __importStar(require("./api"));
__exportStar(require("./accessToken"), exports);
__exportStar(require("./classChannel"), exports);
__exportStar(require("./conferenceDataRetriever"), exports);
__exportStar(require("./conferenceController"), exports);
__exportStar(require("./cryptoAsym"), exports);
__exportStar(require("./endpoints"), exports);
__exportStar(require("./eventSubscriber"), exports);
__exportStar(require("./fullScreenActivityDirector"), exports);
__exportStar(require("./heartbeat"), exports);
__exportStar(require("./ipAddressBlocking"), exports);
__exportStar(require("./limitingRequestHandler"), exports);
__exportStar(require("./logEncryptPublicKey"), exports);
__exportStar(require("./logExporter"), exports);
__exportStar(require("./lsaClient"), exports);
__exportStar(require("./meetingController"), exports);
__exportStar(require("./meetingDataRetriever"), exports);
__exportStar(require("./messageVerification"), exports);
__exportStar(require("./optInParametersStore"), exports);
__exportStar(require("./orgData"), exports);
__exportStar(require("./orgVerificationController"), exports);
__exportStar(require("./pki"), exports);
__exportStar(require("./provisioning"), exports);
__exportStar(require("./socketController"), exports);
__exportStar(require("./socketEndpoint"), exports);
__exportStar(require("./socketIOMessage"), exports);
__exportStar(require("./telemetry"), exports);
__exportStar(require("./utils"), exports);
__exportStar(require("./webHistory"), exports);
__exportStar(require("./shared/state/simpleStateStore"), exports);
