// @lenovo-software/lsa-clients-common/dist/platform.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlatformParser = void 0;
class PlatformParser {
    constructor(osString) {
        this.osString = osString;
    }
    isChromebook() {
        var _a;
        return this.osString === 'Chrome' || ((_a = this.osString) === null || _a === void 0 ? void 0 : _a.includes('CrOS'));
    }
    isWindowsOs() {
        var _a, _b;
        return (_b = (_a = this.osString) === null || _a === void 0 ? void 0 : _a.toLowerCase().includes('windows')) !== null && _b !== void 0 ? _b : false;
    }
    isMacOs() {
        var _a, _b;
        return (_b = (_a = this.osString) === null || _a === void 0 ? void 0 : _a.toLowerCase().includes('mac')) !== null && _b !== void 0 ? _b : false;
    }
    getPlatform() {
        if (this.isChromebook()) {
            return 'Chromebook';
        }
        if (this.isWindowsOs()) {
            return 'Windows';
        }
        if (this.isMacOs()) {
            return 'Mac';
        }
    }
}
exports.PlatformParser = PlatformParser;
